hail/src/main/resources/include/hail/Decoder.h

```diff
@@ -209,6 +211,50 @@ class LEB128InputBuffer {
     bool read_boolean() { return buf_.read_boolean(); }
 };
 
+template<typename Decoder>
+class ReadIterator {
+private:
+  std::shared_ptr<Decoder> dec_;
+  Region * region_;
+  NativeStatus * st_;
+  char * value_;
+
+public:
+  ReadIterator() :
+  dec_(nullptr),
+  region_(nullptr),
+  st_(nullptr),
+  value_(nullptr) { }
+
+  ReadIterator(std::shared_ptr<Decoder> dec, Region * region, NativeStatus* st) :
+  dec_(dec),
+  region_(region),
+  st_(st),
+  value_(nullptr) { }
+
+  ReadIterator<Decoder>& operator++() {
+    if (dec_ != nullptr) {
+      if (dec_->decode_byte(st_)) {
+        value_ = dec_->decode_row(st_, region_);
+      } else {
+        value_ = nullptr;
+        dec_ = nullptr;
+      }
+    }
+    return *this;
+  }
+
+  bool operator==(ReadIterator other) const {
+    return (dec_ == other.dec_) && (value_ == other.value_);
+  }
+
+  bool operator!=(ReadIterator other) const { return !(*this == other); }
+
+  char * operator*() { return value_; }
+  ReadIterator begin() const { return ReadIterator(dec_, region_, st_); }
```

Comment:
I don't think the iterator should have `begin` and `end` methods. Those should go on whatever you're iterating over. As we discussed f2f, let's make a range object that holds the state, and has `begin` and `end` methods, which return iterators that only contain pointers back to the range object.